File: Little Artist/Services/GeminiUsageTracker.swift
//
//  GeminiUsageTracker.swift
//  Little Artist
//
//  Tracks and enforces daily, weekly, and monthly usage limits
//  for Gemini API requests to control cloud AI costs.
//

import Foundation
import os

/// Tracks Gemini API request usage and enforces rate limits.
///
/// Persists counters in UserDefaults with automatic period rollover.
/// Each call to `recordRequest()` increments the current period's count.
/// Call `canMakeRequest()` before every Gemini API call to check limits.
@MainActor
@Observable
final class GeminiUsageTracker {

    // MARK: - Singleton

    static let shared = GeminiUsageTracker()

    // MARK: - Limits

    /// Maximum Gemini requests per day.
    static let dailyLimit = 15

    /// Maximum Gemini requests per week.
    static let weeklyLimit = 60

    /// Maximum Gemini requests per month.
    static let monthlyLimit = 200

    // MARK: - State

    /// Current number of requests made today.
    private(set) var dailyCount: Int = 0

    /// Current number of requests made this week.
    private(set) var weeklyCount: Int = 0

    /// Current number of requests made this month.
    private(set) var monthlyCount: Int = 0

    // MARK: - Persistence Keys

    private enum Keys {
        static let dailyCount = "geminiDailyCount"
        static let weeklyCount = "geminiWeeklyCount"
        static let monthlyCount = "geminiMonthlyCount"
        static let lastDailyReset = "geminiLastDailyReset"
        static let lastWeeklyReset = "geminiLastWeeklyReset"
        static let lastMonthlyReset = "geminiLastMonthlyReset"
    }

    private let defaults = UserDefaults.standard
    private let calendar = Calendar.current
    private let logger = Logger(subsystem: "uk.co.flutterly.Little-Artist", category: "GeminiUsage")

    // MARK: - Init

    private init() {
        rolloverIfNeeded()
    }

    // MARK: - Public API

    /// Whether a new Gemini API request is allowed under all active limits.
    var canMakeRequest: Bool {
        rolloverIfNeeded()
        return dailyCount < Self.dailyLimit
            && weeklyCount < Self.weeklyLimit
            && monthlyCount < Self.monthlyLimit
    }

    /// The most restrictive limit that is currently exceeded, if any.
    var exceededLimit: ExceededLimit? {
        rolloverIfNeeded()
        if dailyCount >= Self.dailyLimit {
            return .daily(resetDate: nextDailyReset)
        }
        if weeklyCount >= Self.weeklyLimit {
            return .weekly(resetDate: nextWeeklyReset)
        }
        if monthlyCount >= Self.monthlyLimit {
            return .monthly(resetDate: nextMonthlyReset)
        }
        return nil
    }

    /// Records a successful Gemini API request.
    func recordRequest() {
        rolloverIfNeeded()
        dailyCount += 1
        weeklyCount += 1
        monthlyCount += 1
        save()
        logger.info("Gemini usage: \(self.dailyCount)/\(Self.dailyLimit) daily, \(self.weeklyCount)/\(Self.weeklyLimit) weekly, \(self.monthlyCount)/\(Self.monthlyLimit) monthly")
    }

    /// Remaining requests for the most restrictive active period.
    var remainingRequests: Int {
        rolloverIfNeeded()
        return min(
            Self.dailyLimit - dailyCount,
            Self.weeklyLimit - weeklyCount,
            Self.monthlyLimit - monthlyCount
        )
    }

    /// Summary of current usage across all periods.
    var usageSummary: UsageSummary {
        rolloverIfNeeded()
        return UsageSummary(
            daily: PeriodUsage(used: dailyCount, limit: Self.dailyLimit, resetDate: nextDailyReset),
            weekly: PeriodUsage(used: weeklyCount, limit: Self.weeklyLimit, resetDate: nextWeeklyReset),
            monthly: PeriodUsage(used: monthlyCount, limit: Self.monthlyLimit, resetDate: nextMonthlyReset)
        )
    }

    // MARK: - Period Rollover

    /// Checks whether any tracking period has elapsed and resets the corresponding counter.
    @discardableResult
    private func rolloverIfNeeded() -> Bool {
        let now = Date()
        var didReset = false

        // Daily reset
        let lastDaily = defaults.object(forKey: Keys.lastDailyReset) as? Date ?? .distantPast
        if !calendar.isDate(lastDaily, inSameDayAs: now) {
            dailyCount = 0
            defaults.set(now, forKey: Keys.lastDailyReset)
            didReset = true
        } else {
            dailyCount = defaults.integer(forKey: Keys.dailyCount)
        }

        // Weekly reset (start of ISO week)
        let lastWeekly = defaults.object(forKey: Keys.lastWeeklyReset) as? Date ?? .distantPast
        let currentWeek = calendar.component(.weekOfYear, from: now)
        let currentWeekYear = calendar.component(.yearForWeekOfYear, from: now)
        let lastWeekNum = calendar.component(.weekOfYear, from: lastWeekly)
        let lastWeekYear = calendar.component(.yearForWeekOfYear, from: lastWeekly)
        if currentWeek != lastWeekNum || currentWeekYear != lastWeekYear {
            weeklyCount = 0
            defaults.set(now, forKey: Keys.lastWeeklyReset)
            didReset = true
        } else {
            weeklyCount = defaults.integer(forKey: Keys.weeklyCount)
        }

        // Monthly reset
        let lastMonthly = defaults.object(forKey: Keys.lastMonthlyReset) as? Date ?? .distantPast
        let currentMonth = calendar.component(.month, from: now)
        let currentYear = calendar.component(.year, from: now)
        let lastMonth = calendar.component(.month, from: lastMonthly)
        let lastYear = calendar.component(.year, from: lastMonthly)
        if currentMonth != lastMonth || currentYear != lastYear {
            monthlyCount = 0
            defaults.set(now, forKey: Keys.lastMonthlyReset)
            didReset = true
        } else {
            monthlyCount = defaults.integer(forKey: Keys.monthlyCount)
        }

        if didReset {
            save()
            logger.info("Gemini usage counters reset")
        }

        return didReset
    }

    // MARK: - Persistence

    private func save() {
        defaults.set(dailyCount, forKey: Keys.dailyCount)
        defaults.set(weeklyCount, forKey: Keys.weeklyCount)
        defaults.set(monthlyCount, forKey: Keys.monthlyCount)
    }

    // MARK: - Reset Date Helpers

    private var nextDailyReset: Date {
        calendar.startOfDay(for: Date()).addingTimeInterval(86400)
    }

    private var nextWeeklyReset: Date {
        var components = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: Date())
        components.weekday = calendar.firstWeekday
        let startOfWeek = calendar.date(from: components) ?? Date()
        return calendar.date(byAdding: .weekOfYear, value: 1, to: startOfWeek) ?? Date()
    }

    private var nextMonthlyReset: Date {
        let components = calendar.dateComponents([.year, .month], from: Date())
        let startOfMonth = calendar.date(from: components) ?? Date()
        return calendar.date(byAdding: .month, value: 1, to: startOfMonth) ?? Date()
    }
}

// MARK: - Supporting Types

/// Describes which usage limit was exceeded and when it resets.
enum ExceededLimit {
    case daily(resetDate: Date)
    case weekly(resetDate: Date)
    case monthly(resetDate: Date)

    var message: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .full
        switch self {
        case .daily(let resetDate):
            return "Daily AI limit reached. Resets \(formatter.localizedString(for: resetDate, relativeTo: Date()))."
        case .weekly(let resetDate):
            return "Weekly AI limit reached. Resets \(formatter.localizedString(for: resetDate, relativeTo: Date()))."
        case .monthly(let resetDate):
            return "Monthly AI limit reached. Resets \(formatter.localizedString(for: resetDate, relativeTo: Date()))."
        }
    }
}

/// Usage data for a single tracking period.
struct PeriodUsage {
    let used: Int
    let limit: Int
    let resetDate: Date

    var remaining: Int { max(0, limit - used) }
    var fraction: Double { Double(used) / Double(limit) }
    var isExceeded: Bool { used >= limit }
}

/// Snapshot of usage across all three periods.
struct UsageSummary {
    let daily: PeriodUsage
    let weekly: PeriodUsage
    let monthly: PeriodUsage
}

File: little_artist_flutter/lib/services/gemini_usage_tracker.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

// ---------------------------------------------------------------------------
// Supporting types
// ---------------------------------------------------------------------------

/// Describes which usage limit was exceeded and when it resets.
enum ExceededLimit {
  daily,
  weekly,
  monthly;

  /// Human-readable message including the relative reset time.
  String message(DateTime resetDate) {
    final now = DateTime.now();
    final diff = resetDate.difference(now);

    String relative;
    if (diff.inDays >= 1) {
      relative = 'in ${diff.inDays} day${diff.inDays == 1 ? '' : 's'}';
    } else if (diff.inHours >= 1) {
      relative = 'in ${diff.inHours} hour${diff.inHours == 1 ? '' : 's'}';
    } else if (diff.inMinutes >= 1) {
      relative = 'in ${diff.inMinutes} minute${diff.inMinutes == 1 ? '' : 's'}';
    } else {
      relative = 'shortly';
    }

    switch (this) {
      case ExceededLimit.daily:
        return 'Daily AI limit reached. Resets $relative.';
      case ExceededLimit.weekly:
        return 'Weekly AI limit reached. Resets $relative.';
      case ExceededLimit.monthly:
        return 'Monthly AI limit reached. Resets $relative.';
    }
  }
}

/// Usage data for a single tracking period.
class PeriodUsage {
  final int used;
  final int limit;
  final DateTime resetDate;

  const PeriodUsage({
    required this.used,
    required this.limit,
    required this.resetDate,
  });

  int get remaining => (limit - used).clamp(0, limit);
  double get fraction => used / limit;
  bool get isExceeded => used >= limit;
}

/// Snapshot of usage across all three periods.
class UsageSummary {
  final PeriodUsage daily;
  final PeriodUsage weekly;
  final PeriodUsage monthly;

  const UsageSummary({
    required this.daily,
    required this.weekly,
    required this.monthly,
  });
}

// ---------------------------------------------------------------------------
// Tracker
// ---------------------------------------------------------------------------

/// Tracks Gemini API request usage and enforces rate limits.
///
/// Persists counters in SharedPreferences with automatic period rollover.
/// Each call to [recordRequest] increments the current period's count.
/// Check [canMakeRequest] before every Gemini API call.
class GeminiUsageTracker extends ChangeNotifier {
  // Limits
  static const dailyLimit = 15;
  static const weeklyLimit = 60;
  static const monthlyLimit = 200;

  // State
  int dailyCount = 0;
  int weeklyCount = 0;
  int monthlyCount = 0;

  // Persistence keys
  static const _keyDailyCount = 'geminiDailyCount';
  static const _keyWeeklyCount = 'geminiWeeklyCount';
  static const _keyMonthlyCount = 'geminiMonthlyCount';
  static const _keyLastDailyReset = 'geminiLastDailyReset';
  static const _keyLastWeeklyReset = 'geminiLastWeeklyReset';
  static const _keyLastMonthlyReset = 'geminiLastMonthlyReset';

  SharedPreferences? _prefs;
  bool _initialised = false;

  /// Loads persisted counters and performs any needed rollovers.
  Future<void> initialise() async {
    if (_initialised) return;
    _prefs = await SharedPreferences.getInstance();
    _rolloverIfNeeded();
    _initialised = true;
  }

  /// Whether a new Gemini API request is allowed under all active limits.
  bool get canMakeRequest {
    _rolloverIfNeeded();
    return dailyCount < dailyLimit &&
        weeklyCount < weeklyLimit &&
        monthlyCount < monthlyLimit;
  }

  /// The most restrictive limit that is currently exceeded, if any.
  ExceededLimit? get exceededLimit {
    _rolloverIfNeeded();
    if (dailyCount >= dailyLimit) return ExceededLimit.daily;
    if (weeklyCount >= weeklyLimit) return ExceededLimit.weekly;
    if (monthlyCount >= monthlyLimit) return ExceededLimit.monthly;
    return null;
  }

  /// The reset date for the currently exceeded limit, if any.
  DateTime? get exceededLimitResetDate {
    _rolloverIfNeeded();
    if (dailyCount >= dailyLimit) return _nextDailyReset;
    if (weeklyCount >= weeklyLimit) return _nextWeeklyReset;
    if (monthlyCount >= monthlyLimit) return _nextMonthlyReset;
    return null;
  }

  /// Records a successful Gemini API request.
  void recordRequest() {
    _rolloverIfNeeded();
    dailyCount++;
    weeklyCount++;
    monthlyCount++;
    _save();
    notifyListeners();
    debugPrint(
      'Gemini usage: $dailyCount/$dailyLimit daily, '
      '$weeklyCount/$weeklyLimit weekly, '
      '$monthlyCount/$monthlyLimit monthly',
    );
  }

  /// Remaining requests for the most restrictive active period.
  int get remainingRequests {
    _rolloverIfNeeded();
    final remaining = [
      dailyLimit - dailyCount,
      weeklyLimit - weeklyCount,
      monthlyLimit - monthlyCount,
    ];
    return remaining.reduce((a, b) => a < b ? a : b);
  }

  /// Summary of current usage across all periods.
  UsageSummary get usageSummary {
    _rolloverIfNeeded();
    return UsageSummary(
      daily: PeriodUsage(
        used: dailyCount,
        limit: dailyLimit,
        resetDate: _nextDailyReset,
      ),
      weekly: PeriodUsage(
        used: weeklyCount,
        limit: weeklyLimit,
        resetDate: _nextWeeklyReset,
      ),
      monthly: PeriodUsage(
        used: monthlyCount,
        limit: monthlyLimit,
        resetDate: _nextMonthlyReset,
      ),
    );
  }

  // ---------------------------------------------------------------------------
  // Period rollover
  // ---------------------------------------------------------------------------

  void _rolloverIfNeeded() {
    final prefs = _prefs;
    if (prefs == null) return;

    final now = DateTime.now();
    bool didReset = false;

    // Daily reset
    final lastDailyMs = prefs.getInt(_keyLastDailyReset);
    final lastDaily = lastDailyMs != null
        ? DateTime.fromMillisecondsSinceEpoch(lastDailyMs)
        : DateTime(2000);
    if (!_isSameDay(lastDaily, now)) {
      dailyCount = 0;
      prefs.setInt(_keyLastDailyReset, now.millisecondsSinceEpoch);
      didReset = true;
    } else {
      dailyCount = prefs.getInt(_keyDailyCount) ?? 0;
    }

    // Weekly reset (different ISO week number or year)
    final lastWeeklyMs = prefs.getInt(_keyLastWeeklyReset);
    final lastWeekly = lastWeeklyMs != null
        ? DateTime.fromMillisecondsSinceEpoch(lastWeeklyMs)
        : DateTime(2000);
    if (_isoWeekNumber(lastWeekly) != _isoWeekNumber(now) ||
        _isoWeekYear(lastWeekly) != _isoWeekYear(now)) {
      weeklyCount = 0;
      prefs.setInt(_keyLastWeeklyReset, now.millisecondsSinceEpoch);
      didReset = true;
    } else {
      weeklyCount = prefs.getInt(_keyWeeklyCount) ?? 0;
    }

    // Monthly reset
    final lastMonthlyMs = prefs.getInt(_keyLastMonthlyReset);
    final lastMonthly = lastMonthlyMs != null
        ? DateTime.fromMillisecondsSinceEpoch(lastMonthlyMs)
        : DateTime(2000);
    if (lastMonthly.month != now.month || lastMonthly.year != now.year) {
      monthlyCount = 0;
      prefs.setInt(_keyLastMonthlyReset, now.millisecondsSinceEpoch);
      didReset = true;
    } else {
      monthlyCount = prefs.getInt(_keyMonthlyCount) ?? 0;
    }

    if (didReset) {
      _save();
      debugPrint('Gemini usage counters reset');
    }
  }

  // ---------------------------------------------------------------------------
  // Persistence
  // ---------------------------------------------------------------------------

  void _save() {
    final prefs = _prefs;
    if (prefs == null) return;
    prefs.setInt(_keyDailyCount, dailyCount);
    prefs.setInt(_keyWeeklyCount, weeklyCount);
    prefs.setInt(_keyMonthlyCount, monthlyCount);
  }

  // ---------------------------------------------------------------------------
  // Reset date helpers
  // ---------------------------------------------------------------------------

  DateTime get _nextDailyReset {
    final now = DateTime.now();
    return DateTime(now.year, now.month, now.day + 1);
  }

  DateTime get _nextWeeklyReset {
    final now = DateTime.now();
    // Monday = 1, Sunday = 7
    final daysUntilNextMonday = (DateTime.monday - now.weekday + 7) % 7;
    final nextMonday =
        daysUntilNextMonday == 0 ? 7 : daysUntilNextMonday;
    return DateTime(now.year, now.month, now.day + nextMonday);
  }

  DateTime get _nextMonthlyReset {
    final now = DateTime.now();
    return DateTime(now.year, now.month + 1, 1);
  }

  // ---------------------------------------------------------------------------
  // Date helpers
  // ---------------------------------------------------------------------------

  static bool _isSameDay(DateTime a, DateTime b) {
    return a.year == b.year && a.month == b.month && a.day == b.day;
  }

  /// ISO 8601 week number.
  static int _isoWeekNumber(DateTime date) {
    // Algorithm: find the Thursday of this date's week, then count weeks.
    final thursday = date.add(Duration(days: DateTime.thursday - date.weekday));
    final jan1 = DateTime(thursday.year, 1, 1);
    return ((thursday.difference(jan1).inDays) / 7).ceil() + 1;
  }

  /// ISO 8601 week-numbering year.
  static int _isoWeekYear(DateTime date) {
    final thursday = date.add(Duration(days: DateTime.thursday - date.weekday));
    return thursday.year;
  }
}
